backend/attachRendererFiber.js

```diff
@@ -0,0 +1,244 @@
+      });
+    }
+  }
+
+  function enqueueUpdate(fiber, events) {
+    events.push({
+      element: fiber._debugID,
+      data: getDataFiber(fiber),
+      renderer: rid,
+      _event: 'update',
+    });
+  }
+
+  function enqueueUnmount(fiber, events) {
+    events.push({
+      element: fiber._debugID,
+      renderer: rid,
+      _event: 'unmount',
+    });
+  }
+
+  function mapChildren(parent, allKeys) {
+    const children = new Map();
+    let node = parent.child;
+    while (node) {
+      const key = node.key || node.index;
+      allKeys.add(key);
+      children.set(key, node);
+      node = node.sibling;
+    }
+    return children;
+  }
+
+  function unmountFiber(fiber, events) {
+    // Depth-first.
+    // Logs unmounting of children first, parents later.
+    let node = fiber;
+    outer: while (true) {
+      if (node.child) {
+        node.child.return = node;
+        node = node.child;
+        continue;
+      }
+      enqueueUnmount(node, events);
+      if (node == fiber) {
+        return;
+      }
+      if (node.sibling) {
+        node.sibling.return = node.return;
+        node = node.sibling;
+        continue;
+      }
+      while (node.return) {
+        node = node.return;
+        enqueueUnmount(node, events);
+        if (node == fiber) {
+          return;
+        }
+        if (node.sibling) {
+          node.sibling.return = node.return;
+          node = node.sibling;
+          continue outer;
+        }
+      }
+      return;
+    }
+  }
+
+  function mountFiber(fiber, events) {
+    // Depth-first.
+    // Logs mounting of children first, parents later.
+    let node = fiber;
+    outer: while (true) {
+      if (node.child) {
+        node.child.return = node;
+        node = node.child;
+        continue;
+      }
+      enqueueMount(node, events);
+      if (node == fiber) {
+        return;
+      }
+      if (node.sibling) {
+        node.sibling.return = node.return;
+        node = node.sibling;
+        continue;
+      }
+      while (node.return) {
+        node = node.return;
+        enqueueMount(node, events);
+        if (node == fiber) {
+          return;
+        }
+        if (node.sibling) {
+          node.sibling.return = node.return;
+          node = node.sibling;
+          continue outer;
+        }
+      }
+      return;
+    }
+  }
+
+  function updateFiber(nextFiber, prevFiber, events) {
+    // TODO: optimize for the common case of children with implcit keys.
+    // Just like we do in the Fiber child reconciler.
+    // We shouldn't be allocating Maps all the time.
+    const allKeys = new Set();
+    const prevChildren = mapChildren(prevFiber, allKeys);
+    const nextChildren = mapChildren(nextFiber, allKeys);
+    allKeys.forEach(key => {
+      const prevChild = prevChildren.get(key);
+      const nextChild = nextChildren.get(key);
+      if (prevChild != null && nextChild != null) {
+        // TODO: are there more cases when we can bail out?
```

Comment:
Currently, either all children are cloned or none are cloned. So you really only need to check `nextFiber.child === prevFiber.child` to know if all are the same or all are different.